utils/optiwebp.py

```diff
@@ -0,0 +1,160 @@
+            elif filetype == ".webp":
+                with open(os.path.join(image_root, filename), "rb") as img:
+                    # discard first 12 bytes of WebP "magic number" and file size
+                    img.read(12)
+                    # read type of WebP file
+                    # VP8L = lossless
+                    # VP8 = lossy
+                    # VP8X = can either, so assume lossless (better to try to go from lossy->lossless than the reverse)
+                    webp_type = img.read(4)
+                    if webp_type == b"VP8L" or webp_type == b"VP8X":
+                        subprocess_args = lossless_args
+                    elif webp_type == b"VP8 ":
+                        subprocess_args = lossy_args
+                    else:
+                        print(filename, "is not a valid WebP file", file=sys.stderr)
+                        continue
+            elif filetype == ".jpg":
+                subprocess_args = lossy_args
+            else:
+                continue
+
+            initial_file = os.path.join(image_root, filename)
+            webp_file = ""
+            if filetype != ".webp":
+                webp_file = os.path.join(options.tempdir, filename)[:-3]+"webp"
+            else:
+                webp_file = os.path.join(options.tempdir, filename)
+
+            result = subprocess.run(subprocess_args + [initial_file, "-o", webp_file], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+            if result.returncode != 0:
+                print(initial_file+" failed conversion to webp.")
+                continue
+
+            initial_size = os.path.getsize(initial_file)
+            converted_size = os.path.getsize(webp_file)
+            percentage_change = round((converted_size/initial_size)*100, 2)
+            absolute_change = initial_size - converted_size
+
+            results_dict[initial_file] = {
+                "type": filetype,
+                "new_file": webp_file,
+                "old_size": initial_size,
+                "new_size": converted_size,
+                "change_in_bytes": absolute_change,
+                "change_in_percent": percentage_change
+            }
+
+            if options.nodryrun and percentage_change < options.minpercent:
+                os.remove(initial_file)
+                shutil.copy(webp_file, image_root)
+
+            os.remove(webp_file)
+
+good_conversion = "good-conversion.tsv"
+bad_conversion = "bad-conversion.tsv"
+
+if os.path.exists(good_conversion):
+    os.remove(good_conversion)
+with open(good_conversion, "a") as f:
+    f.write("filename,old_size,new_size,change_in_percent,change_in_bytes\n")
+
+if os.path.exists(bad_conversion):
+    os.remove(bad_conversion)
+with open(bad_conversion, "a") as f:
+    f.write("filename,old_size,new_size,change_in_percent,change_in_bytes\n")
+
+initial_total_size = 0
+final_total_size = 0
+
+with open(good_conversion, "a") as good_f, open(bad_conversion, "a") as bad_f:
```

Comment:
If you move lines 129 and 134 here, you can replace the opening mode `"a"` with `"w"` and overwrite the files even if they already exist (no need to remove a file then append to an empty file).
You should also wrap this block inside a `try ... except OSError:` block, in case the user cannot write to the files for any reason (like missing permissions).

Reply:
Done.